receiver/mongodbreceiver/extractor.go

```diff
@@ -0,0 +1,437 @@
+	switch m.dataPointType {
+	case integer:
+		e.addIntMetric(ts, mb, document, m, dbName)
+	case double:
+		e.addDoubleMetric(ts, mb, document, m, dbName)
+	default:
+		e.logger.Debug("unknown data point type", zap.String("name", m.name))
+	}
+}
+
+func (e *extractor) addIntMetric(ts pdata.Timestamp, mb *metadata.MetricsBuilder, document bson.M, metric mongoMetric, dbName string) {
+	value, err := digForIntValue(document, metric.path)
+	if err != nil {
+		e.logger.Debug("unable to find metric", zap.String("name", metric.name))
+		return
+	}
+
+	if metric.conversionFactor != 0 {
+		value = metric.convertToAppropriateValue(value).(int64)
+	}
+
+	switch metric.name {
+	// mongodb.collection.count
+	case collectionCount:
+		mb.RecordMongodbCollectionCountDataPoint(ts, value, dbName)
+	// mongodb.connection.count
+	case connections:
+		mb.RecordMongodbConnectionCountDataPoint(ts, value, dbName, metric.staticAttributes[metadata.A.ConnectionType])
+	// mongodb.data.size
+	case dataSize:
+		mb.RecordMongodbDataSizeDataPoint(ts, value, dbName)
+	// mongodb.index.count
+	case indexes:
+		mb.RecordMongodbIndexCountDataPoint(ts, value, dbName)
+	// mongodb.index.size
+	case indexSize:
+		mb.RecordMongodbIndexSizeDataPoint(ts, value, dbName)
+	// mongodb.memory.usage
+	case memUsage:
+		mb.RecordMongodbMemoryUsageDataPoint(ts, value, dbName, metric.staticAttributes[metadata.A.MemoryType])
+	// mongodb.extent.count
+	case numExtents:
+		mb.RecordMongodbExtentCountDataPoint(ts, value, dbName)
+	// mongodb.object.count
+	case objects:
+		mb.RecordMongodbObjectCountDataPoint(ts, value, dbName)
+	// mongodb.storage.size
+	case storageSize:
+		mb.RecordMongodbStorageSizeDataPoint(ts, value, dbName)
+	default:
+		e.logger.Warn("attempted to add an unknown metric", zap.String("metric", metric.name))
+	}
+}
+
+func (e *extractor) addDoubleMetric(_ pdata.Timestamp, _ *metadata.MetricsBuilder, document bson.M, metric mongoMetric, _ string) {
+	_, err := digForDoubleValue(document, metric.path)
+	if err != nil {
+		e.logger.Debug("unable to find metric", zap.String("name", metric.name))
+		return
+	}
+
+	switch metric.name {
+	default:
+		e.logger.Warn("attempted to add metric datapoint was collected.", zap.String("name", metric.name))
+	}
+}
+
+func (e *extractor) shouldDig(metric mongoMetric) bool {
+	if metric.maxMongoVersion == nil && metric.minMongoVersion == nil {
+		return true
+	}
+
+	satisfyLowerBound, satisfyUpperBound := true, true
+	if metric.minMongoVersion != nil {
+		satisfyLowerBound = e.version.GreaterThanOrEqual(metric.minMongoVersion)
+	}
+
+	if metric.maxMongoVersion != nil {
+		satisfyUpperBound = e.version.LessThanOrEqual(metric.maxMongoVersion)
+	}
+
+	return satisfyLowerBound && satisfyUpperBound
+}
+
+func digForIntValue(document bson.M, path []string) (int64, error) {
+	curItem, remainingPath := path[0], path[1:]
+	value := document[curItem]
+	if value == nil {
+		return 0, errors.New("nil found when digging for metric")
+	}
```

Comment:
Looks like we got rid of the "double" version of this. That's fine, but it's still doing two unrelated things. I suggest splitting into "dig" and "parse" functions, and calling both of them at the same level so that errors can be handled appropriately.

```
v, ok := dig(data, path)
if !ok // partial error, skip parse

p, err := parse(v)
if err != nil // partial error, skip record

mb.recordFoo(p)
```